### DISCUSSION TITLE
cargo-auditable and cargo-chef or crate2nix

### DISCUSSION THREAD

**Discussion by lfrancke** [rust-secure-code/cargo-auditable#101] [Q&A] [answered]
Hi,

are you aware whether cargo-auditable should work with [cargo-chef](https://github.com/LukeMathWalker/cargo-chef) or [crate2nix](https://github.com/kolloch/crate2nix)?
We plan on testing that this week but I thought I could maybe draw on previous experience.

All of these tools somehow wrap cargo but I'm not too sure how much they play around with the internals.

Thanks!

**Comment by bjorn3** [comment]
Crate2nix probably isn't going to work as cargo-auditable expects the inner cargo invocation to directly compile the program and passes a couple of env vars to it's rustc wrapper. I expect crate2nix to miss these env vars and possibly even the rustc wrapper. As for cargo-chef, if you can tell cargo-chef to invoke `cargo auditable build` instead of `cargo build` inside the docker container (and make sure to install `cargo-auditable` in the container) I would expect it to work.

**Reply by Shnatsel** [reply]
Or you can make `cargo auditable` a drop-in replacement for Cargo within the container as described [here](https://github.com/rust-secure-code/cargo-auditable/blob/master/REPLACING_CARGO.md), and then you won't have to modify what `cargo-chef` invokes at all.

**Comment by Shnatsel** [comment]
@figsoda might be answer the bit about crate2nix, since he's been working on enabling `cargo auditable` in NixOS by default.

**Comment by figsoda** [comment]
I am not very familiar with `buildRustCrate` (which crate2nix uses under the hood), but looks like it mostly uses rustc under the hood and doesn't work with `cargo-auditable`

I tried to build crate2nix with cargo-auditable with no success, here is the WIP patch (doesn't work) that tries to emulate what [cargo_auditable.rs](https://github.com/rust-secure-code/cargo-auditable/blob/master/cargo-auditable/src/cargo_auditable.rs) does

```patch
--- a/default.nix
+++ b/default.nix
@@ -12,7 +12,19 @@
 , release ? true
 }:
 let
-  cargoNix = callPackage ./crate2nix/Cargo.nix { inherit release; };
+  cargoNix = callPackage ./crate2nix/Cargo.nix {
+    inherit release;
+    buildRustCrateForPkgs = pkgs: pkgs.buildRustCrate.override {
+      cargo = pkgs.cargo-auditable-cargo-wrapper;
+      rustc = pkgs.writeShellScriptBin "rustc" ''
+        export CARGO=${pkgs.cargo}/bin/cargo
+        export CARGO_AUDITABLE_ORIG_ARGS=[]
+        export PATH=${pkgs.rustc}/bin:$PATH
+        export RUSTC_WORKSPACE_WRAPPER=${pkgs.cargo-auditable}/bin/cargo-auditable
+        ${pkgs.cargo-auditable}/bin/cargo-auditable rustc "$@"
+      '';
+    };
+  };
   withoutTemplates = name: type:
     let
       baseName = builtins.baseNameOf (builtins.toString name);
--- a/nix/sources.json
+++ b/nix/sources.json
@@ -53,10 +53,10 @@
         "homepage": "https://github.com/NixOS/nixpkgs",
         "owner": "NixOS",
         "repo": "nixpkgs",
-        "rev": "5f588eb4a958f1a526ed8da02d6ea1bea0047b9f",
-        "sha256": "1wgk7d1j8yj3arsbiva69bicnf3m9id8knjp625knq0siv3i9xsg",
+        "rev": "5a350a8f31bb7ef0c6e79aea3795a890cf7743d4",
+        "sha256": "0fjii4ay8lp1byq74vq6p5bv754ssbf0ij5xwr5zr27db23ws77j",
         "type": "tarball",
-        "url": "https://github.com/NixOS/nixpkgs/archive/5f588eb4a958f1a526ed8da02d6ea1bea0047b9f.tar.gz",
+        "url": "https://github.com/NixOS/nixpkgs/archive/5a350a8f31bb7ef0c6e79aea3795a890cf7743d4.tar.gz",
         "url_template": "https://github.com/<owner>/<repo>/archive/<rev>.tar.gz"
     },
     "nixpkgs-mozilla": {
```

**Reply by Shnatsel** [reply]
Note that since this is an emulation, it relies on non-public APIs that are not guaranteed to be stable across versions. So if you use this approach, you'll need to pin a specific version of `cargo-auditable`.

**Reply by figsoda** [reply]
In case this is not clear from the original text, it also doesn't even work and was just an attempt

**Reply by lfrancke** [reply]
I've yet to use Nix, but it's fairly easy to read and I see what you're trying.
What was the problem?

**Reply by figsoda** [reply]
The problem is that `cargo-auditable` works with `cargo`, but `crate2nix` uses `rustc` directly (aside from a couple `cargo metadata` calls)

**Comment by lfrancke** [comment]
Wow! Thank you so much everyone for the detailed (and fast) responses :bow: 
I will dig through them later this week and we'll decide on how to go forward. I'll update here.

**Reply by lfrancke** [reply, marked as answer]
We decided that (for now) we'll use two builds: For release we're not going to use crate2nix/cargo-chef but normal cargo + auditable and for local dev we're going to use one of those solutions.

Thanks again!
